server/task_scheduler.go

```diff
@@ -149,18 +151,54 @@ func (s *TaskScheduler) Run(ctx context.Context, wg *sync.WaitGroup) {
 						)
 						continue
 					}
+
+					executor := executorGetter()
 					s.runningExecutorsMutex.Lock()
-					s.runningExecutors[task.ID] = executorGetter()
+					s.runningExecutors[task.ID] = executor
 					s.runningExecutorsMutex.Unlock()
 
-					go func(task *api.Task, executor TaskExecutor) {
+					go func(ctx context.Context, task *api.Task, executor TaskExecutor) {
 						defer func() {
 							s.runningExecutorsMutex.Lock()
 							delete(s.runningExecutors, task.ID)
+							delete(s.runningExecutorsCancel, task.ID)
 							s.runningExecutorsMutex.Unlock()
 							s.taskProgress.Delete(task.ID)
 						}()
-						done, result, err := RunTaskExecutorOnce(ctx, executor, s.server, task)
+
+						executorCtx, cancel := context.WithCancel(ctx)
+						s.runningExecutorsMutex.Lock()
```

Comment:
If there is a lot of code we call `Unlock()` directly instead of using `defer Unlock()`, can we wrap the `runningExecutors `and provide `cancel()`, `delete()` function to caller?

Reply:
Sure. Let's do that when there is more duplication.